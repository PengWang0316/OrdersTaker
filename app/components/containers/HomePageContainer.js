import React, { Component, Fragment } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

import { fetchAllMenu } from '../../actions/MenuActions';

import Banner from '../Banner';
import MenuCategory from '../MenuCategory';
import ItemDetailDialog from '../ItemDetailDialog';
import ShowDetailDialogContext from '../../contexts/ShowDetailDialogContext'; // Import the context to pass the function.


/** The component that shows the home page containt */
export class HomePageContainer extends Component {
  static propTypes = {
    menus: PropTypes.array,
    fetchAllMenu: PropTypes.func.isRequired
  };
  static defaultProps = { menus: null };

  /**
   * call fetchAllMenu method when the menus is null
   * @param {object} props is an object that contains props' values
   */
  constructor(props) {
    super(props);
    if (!props.menus) props.fetchAllMenu(); // Fetching the menu data if it has not exsited in the Redux yet.
    window.addEventListener('resize', this.handleResize);
  }

  /**
   * itemAmount is the number will be showed on the page (Rest of them will be hidden in an expaned panel.)
   * width > 1200 shows 12, > 987 shows 10, >801 shows 8, < 615 shows 6
   */
  state = {
    itemAmount: HomePageContainer.getAmountNumber(window.innerWidth),
    isDialogOpen: false,
    currentItem: null
  };


  /**
   * Calculating the total amount of item will show.
   * @param {number} width is the current browser width.
   * @return {number} Return a number will be showed. width > 1200 shows 12, > 987 shows 10, >801 shows 8, < 615 shows 6
   */
  static getAmountNumber = width => {
    if (width <= 801) return 6;
    else if (width <= 987) return 8;
    else if (width <= 1200) return 10;
    return 12;
  };

  /**
   * Using to get current browser width and set it to state.
   * @return {null} No return.
   */
  handleResize = () => this.setState({
    itemAmount: HomePageContainer.getAmountNumber(window.innerWidth)
  });

  /**
   * Showing the item detail dialog.
   * Making sure the currentItem has been setup before open the dialog.
   * @param {string} itemId is the id of an item.
   * @return {null} No return.
   */
  showDetailDialog = itemId => this.setState(
    { currentItem: this.props.menuItems[itemId] },
    () => this.handleDialogToggle()
  );

  /**
   * Setting the isDialogOpen state to a opposite value.
   * @return {null} Ignore tht return.
   */
  handleDialogToggle = () => this.setState(({ isDialogOpen }) => ({ isDialogOpen: !isDialogOpen }));

  /**
   * Rendering the jsx for the component.
   * @return {jsx} Return jsx
   */
  render() {
    const { menus } = this.props;
    const { itemAmount, isDialogOpen, currentItem } = this.state;
    return (
      <Fragment>
        <Banner />
        <ShowDetailDialogContext.Provider value={this.showDetailDialog}>
          {menus && menus.map(menu =>
            <MenuCategory menu={menu} key={menu._id} itemAmount={itemAmount} />)}
        </ShowDetailDialogContext.Provider>
        <ItemDetailDialog onClose={this.handleDialogToggle} open={isDialogOpen} item={currentItem} />
      </Fragment>
    );
  }
}

/* istanbul ignore next */
const mapStateToProps = state => ({
  menus: state.menus,
  menuItems: state.menuItems
});
/* istanbul ignore next */
const mapDispatchToProps = dispatch => ({
  fetchAllMenu: () => dispatch(fetchAllMenu())
});
export default connect(mapStateToProps, mapDispatchToProps)(HomePageContainer);
