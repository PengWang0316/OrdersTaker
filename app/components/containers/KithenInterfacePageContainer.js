import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { Typography } from '@material-ui/core';
import { withStyles } from '@material-ui/core/styles';
import Clock from 'react-live-clock';

import { SUPER_USER_ROLE, HOME_PAGE_URL } from '../../config';
import KithenOrderBoard from '../KithenOrderBoard';
import { fetchAllMenu } from '../../actions/MenuActions';

const styles = {
  root: {
    width: '95%',
    height: '100%',
    margin: '20px auto'
  },
  titleDiv: {
    display: 'flex',
    justifyContent: 'space-between',
    padding: 10
  }
};

/**
 * The component shows an interface for the kithen.
 * It will allow people who work in the kithen to see and interact with the orders that customers made.
 */
export class KithenInterfacePageContainer extends Component {
  static propTypes = {
    classes: PropTypes.object.isRequired,
    fetchAllMenu: PropTypes.func.isRequired,
    user: PropTypes.object.isRequired,
    menuItems: PropTypes.object.isRequired
  };

  /**
   * Redirecting the user to the home page if the user does not have an approperate role.
   * @param {object} props has all component's prop value.
   * @return {null} No return.
   */
  constructor(props) {
    super(props);
    const { user, history } = props;
    if (!user._id || !user.role || user.role < SUPER_USER_ROLE) {
      history.push(HOME_PAGE_URL);
      return;
    }
    if (Object.keys(props.menuItems).length === 0) props.fetchAllMenu();
    this.timezone = Intl.DateTimeFormat().resolvedOptions().timeZone; // Saving the timezone name
  }

  /**
   * Render method.
   * @return {jsx} Return jsx.
   */
  render() {
    const { classes } = this.props;
    return (
      <div className={classes.root}>
        <div className={classes.titleDiv}>
          <Typography color="primary" variant="title">All received order</Typography>
          <Typography color="textSecondary">
            <Clock format="HH:mm:ss MMM DD YYYY" ticking timezone={this.timezone} />
          </Typography>
        </div>
        <KithenOrderBoard />
      </div>
    );
  }
}
/* istanbul ignore next */
const mapStateToProps = state => ({
  menuItems: state.menuItems,
  user: state.user
});
/* istanbul ignore next */
const mapDispatchToProps = dispatch => ({
  fetchAllMenu: () => dispatch(fetchAllMenu())
});
export default connect(mapStateToProps, mapDispatchToProps)(withStyles(styles)(KithenInterfacePageContainer));
