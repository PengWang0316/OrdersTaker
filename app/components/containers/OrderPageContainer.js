import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles';

import OrderSummary from '../OrderSummary';
import OrderList from '../OrderList';
import ItemDetailDialog from '../ItemDetailDialog';
import { fetchAllMenu } from '../../actions/MenuActions';

const styles = {
  root: {
    maxWidth: 1200,
    padding: '10px 20px',
    margin: '15px auto',
    display: 'grid',
    gridTemplateColumns: '1fr 200px',
    gridTemplateRows: '1fr',
    gridTemplateAreas: '"listPanel summaryPanel"',
    gridGap: '2em'
  },
  listPanel: {
    gridArea: 'listPanel'
  },
  summaryPanel: {
    gridArea: 'summaryPanel'
  },
  summaryContent: {
    position: 'sticky',
    top: 70
  },
  '@media (max-width: 600px)': {
    root: {
      gridTemplateColumns: '1fr',
      gridTemplateRows: '1fr 1fr',
      gridTemplateAreas: `"summaryPanel"
                          "listPanel"`
    }
  }
};

/**
 * Showing the customers' order detail and allow them to add number for items, remove items, clear orders, scan QR code for the table, and place orders.
 */
export class OrderPageContainer extends Component {

  /**
   * Parsing the Redux state orders to an object this component can use. Also calculate the price for the order.
   * @param {object} orders comes from Redux's state.
   * @return {object} Return an object that is sorted to use category to be the key.
   * The return object structure is like below:
   * {
   *    categories: {
   *      categoryA: { price: xx, qty: xx, ids: [id1, id2, ...] },
   *      ...
   *    },
   *    totalQty: xxx,
   *    price: xxx,
   *    tax: xxx,
   *    totalPrice: xxx,
   * }
   */
  static parseOrders = orders => {
    const newOrders = { // Initial some properties
      totalQty: 0, price: 0, tax: 0, totalPrice: 0, categories: {}
    };
    // price and tax will use cent in order to avoid the inaccurace.
    Object.keys(orders).forEach(key => {
      if (key === 'qty') return;
      const category = newOrders.categories[orders[key].category];
      Object.keys(orders[key].qty).forEach(priceKey => { // Adding all price
        const price = orders[key].prices[priceKey];
        const qty = orders[key].qty[priceKey];
        // Calculating the value for a specific category
        newOrders.categories[orders[key].category] = category ?
          { price: ((category.price * 100) + (price * 100 * qty)) / 100, qty: category.qty + qty, ids: category.ids } :
          { price: (price * 100 * qty) / 100, qty, ids: [] }; // If the category is empty, initialize a empty ids array.
        newOrders.categories[orders[key].category].ids.push(key); // Saving the id to the ids field.
        // Calculating the total value
        newOrders.totalQty += qty; // Adding the total quantity
        newOrders.price = ((newOrders.price * 100) + (price * 100 * qty)) / 100; // Adding the price up
        newOrders.tax = (((newOrders.tax * 100) + (((orders[key].taxRate * 100) * (price * 100 * qty)) / 10000)) / 100).toFixed(2);
      });
    });
    newOrders.totalPrice = ((newOrders.price * 100) + (newOrders.tax * 100)) / 100;
    return newOrders;
  };

  static propTypes = {
    menuItems: PropTypes.object.isRequired,
    orders: PropTypes.object.isRequired,
    classes: PropTypes.object.isRequired,
    fetchAllMenu: PropTypes.func.isRequired
  };

  /**
   * Call the fetchAllMenu to get Redux state for its children components if the menus is still null.
   * @param {object} props contains the value for the componnet.
   * @return {null} No return.
   */
  constructor(props) {
    super(props);
    if (Object.keys(props.menuItems).length === 0) props.fetchAllMenu();
  }

  state = {
    isDialogOpen: false,
    currentItem: null
  }

  /**
   * Showing the item detail dialog.
   * Making sure the currentItem has been setup before open the dialog.
   * @param {string} itemId is the id of an item.
   * @return {null} No return.
   */
  showDetailDialog = itemId => this.setState(
    { currentItem: this.props.menuItems[itemId] },
    () => this.handleDialogToggle()
  );

  /**
   * Setting the isDialogOpen state to a opposite value.
   * @return {null} Ignore tht return.
   */
  handleDialogToggle = () => this.setState(({ isDialogOpen }) => ({ isDialogOpen: !isDialogOpen }));

  /**
   * The render method
   * @return {jsx} Return the jsx for the component.
   */
  render() {
    const { classes, menuItems, orders } = this.props;
    const { currentItem, isDialogOpen } = this.state;
    const newOrders = OrderPageContainer.parseOrders(orders);
    return (
      <div className={classes.root}>
        <div className={classes.summaryPanel}>
          <div className={classes.summaryContent}><OrderSummary orders={newOrders} /></div>
        </div>
        <div className={classes.listPanel}>
          <OrderList orders={newOrders} />
        </div>
        {Object.keys(menuItems).length === 0 && <ItemDetailDialog onClose={this.handleDialogToggle} open={isDialogOpen} item={currentItem} />}
      </div>
    );
  }
}
/* istanbul ignore next */
const mapStateToProps = state => ({
  menuItems: state.menuItems,
  orders: state.orders
});
/* istanbul ignore next */
const mapDispatchToProps = dispatch => ({
  fetchAllMenu: () => dispatch(fetchAllMenu())
});
export default connect(mapStateToProps, mapDispatchToProps)(withStyles(styles)(OrderPageContainer));
