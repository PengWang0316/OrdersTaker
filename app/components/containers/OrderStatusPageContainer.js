import React, { Component } from 'react';

/**
 * Show the detailed information about an order's status.
 */
export class OrderStatusPageContainer extends Component {
  /**
   * Load the data based on the order id the component retreived from the url.
   * @param {object} props contains all props' value for the component.
   * @return {null} No return.
   */
  constructor(props) {
    super(props);
    this.state = {
      orderId: props.match.params.orderId
    };
  }

  /**
   * The render method for the component.
   * @return {jsx} Return the jsx.
   */
  render() {
    const { orderId } = this.state;
    return (
      <div>{orderId}</div>
    );
  }
}
export default OrderStatusPageContainer;
