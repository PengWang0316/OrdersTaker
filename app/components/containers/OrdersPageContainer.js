import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import { fetchOrderAmount, fetchUnloginUserOrders, fetchLoginUserOrders } from '../../actions/OrdersActions';
import { removeTempOrderId, linkOrderToAccount } from '../../actions/TempOrderIdsActions';
import { fetchAllMenu } from '../../actions/MenuActions';
import OrderBrowser from '../OrderBrowser';
import OrdersPageContext from '../../contexts/OrdersPageContext';
import OrderDetailDialog from '../OrderDetailDialog';
import AlertDialog from '../AlertDialog';
import CheckCircleIconSnackbar from '../snackbars/CheckCircleIconSnackbar';

/**
 * The component shows a user's order records.
 * This component will show both login user records and the orders were saved to Redux state when the user was not login.
 */
export class OrdersPageContainer extends Component {
  static propTypes = {
    user: PropTypes.object.isRequired,
    orders: PropTypes.object.isRequired,
    tempOrderIds: PropTypes.array.isRequired,
    menuItems: PropTypes.object.isRequired,
    fetchOrderAmount: PropTypes.func.isRequired,
    fetchAllMenu: PropTypes.func.isRequired,
    loginUserOrderAmount: PropTypes.number,
    removeTempOrderId: PropTypes.func.isRequired
  };

  static defaultProps = {
    loginUserOrderAmount: null
  };

  /**
   * Call the fetchAllMenu to get Redux state for its children components if the menus is still null.
   * @param {object} props contains the value for the componnet.
   * @return {null} No return.
   */
  constructor(props) {
    super(props);
    if (Object.keys(props.menuItems).length === 0) props.fetchAllMenu();
  }

  state = {
    isDetailedDialogOpen: false, // The indicator of showing the detailed order dialog.
    isRemoveAlertOpen: false, // The indicator of showing the remove alert dialog.
    isLinkAlertOpen: false, // The indicator of showing the link alert dialog.
    isSnackbarOpen: false,
    snackbarMessage: '',
    // loginUserOrderOffset: 0,
    // unLoginUserOrderOffset: 0,
    loginUserOrders: [],
    unloginUserOrders: []
  }

  /**
   * Calling the fetchOrderAmount if the user has already logged in and the amount is still be null.
   * @return {null} No return.
   */
  componentDidMount() {
    if (this.props.user._id) {
      this.fetchLoginUserOrders(0);
      if (this.props.orders.loginUserOrderAmount === null) this.props.fetchOrderAmount(this.props.user);
    }
    if (this.props.tempOrderIds.length !== 0) this.fetchUnloginUserOrders(0);
  }

  /**
   * Removing the element has _id equal deleteOrderId
   * @param {array} unloginUserOrders is the old order array.
   * @param {string} deleteOrderId is the order id will be removed from the original array.
   * @return {array} Return an array without the element that has _id equal deleteOrderId.
   */
  static getNewUnloginOrdersArray = (unloginUserOrders, deleteOrderId) => {
    const newArray = [];
    unloginUserOrders.forEach(element => { if (element._id !== deleteOrderId) newArray.push(element); });
    return newArray;
  };

  /**
   * Fetching the orders for a login user.
   * @param {int} offset is the how many order will be skipped.
   * @return {null} No return.
   */
  fetchLoginUserOrders = offset => fetchLoginUserOrders(offset, this.props.user)
    .then(loginUserOrders => this.setState({ loginUserOrders })).catch(err => console.error(err));

  /**
   * Fetching the orders for a unlogin user.
   * @param {int} offset is the how many order will be skipped.
   * @return {null} No return.
   */
  fetchUnloginUserOrders = offset => fetchUnloginUserOrders(offset, this.props.tempOrderIds)
    .then(unloginUserOrders => this.setState({ unloginUserOrders })).catch(err => console.error(err));

  /**
   * Toggling the isOpenDetailDialog state.
   * @return {null} No return.
   */
  toggleDetailDialog = () => this.setState(({ isDetailedDialogOpen }) => ({ isDetailedDialogOpen: !isDetailedDialogOpen }));

  /**
   * Setting up the order to currentOrder state and call toggleDetailDialog function.
   * @param {object} order is the current order will be showed in the detailed dailog
   * @return {null} No return
   */
  handleShowDetail = order => this.setState({ currentOrder: order }, () => this.toggleDetailDialog());

  /**
   * Showing the confirm dailog and handle the remove process
   * @return {null} No return
   */
  handleRmoveUnloginOrder = () => {
    const orderId = this.unloginOrderId;
    this.props.removeTempOrderId(this.unloginOrderId);
    this.handleToggleRemoveAlertDialog(null);
    return this.setState(({ unloginUserOrders }) => ({ unloginUserOrders: OrdersPageContainer.getNewUnloginOrdersArray(unloginUserOrders, orderId), snackbarMessage: 'Removed Successfully' }), () => this.handleToggleSnackbar());
  }

  /**
   * Showing a confirm dailog and handle the linking process
   * @return {null} No return
   */
  handleLinkOrder = () => {
    const orderId = this.unloginOrderId;
    this.props.linkOrderToAccount(orderId, this.props.user.jwt);
    this.handleToggleLinkAlertDialog(null);
    return this.setState(({ unloginUserOrders }) => ({ unloginUserOrders: OrdersPageContainer.getNewUnloginOrdersArray(unloginUserOrders, orderId), snackbarMessage: 'Linked to your account sccessfully' }), () => this.handleToggleSnackbar());
  };

  /**
   * Setting the isLinkAlertOpen state to the opposite value.
   * @param {string} orderId is the id of the unlogin order.
   * @return {null} No return.
   */
  handleToggleRemoveAlertDialog = orderId => {
    this.unloginOrderId = orderId;
    return this.setState(({ isRemoveAlertOpen }) => ({ isRemoveAlertOpen: !isRemoveAlertOpen }));
  }

  /**
   * Setting the isRemoveAlertOpen state to the opposite value.
   * @param {string} orderId is the id of the unlogin order.
   * @return {null} No return.
   */
  handleToggleLinkAlertDialog = orderId => {
    this.unloginOrderId = orderId;
    return this.setState(({ isLinkAlertOpen }) => ({ isLinkAlertOpen: !isLinkAlertOpen }));
  }

  /**
   * Setting up the state isSnackbarOpen to an opposite value.
   * @return {null} No return.
   */
  handleToggleSnackbar = () => this.setState(({ isSnackbarOpen }) => ({ isSnackbarOpen: !isSnackbarOpen }));

  /**
   * The render method
   * @return {jsx} Return the jsx for the component
   */
  render() {
    const { tempOrderIds, loginUserOrderAmount, menuItems, user } = this.props;
    const { loginUserOrders, unloginUserOrders, currentOrder, isDetailedDialogOpen, isRemoveAlertOpen, isLinkAlertOpen, isSnackbarOpen, snackbarMessage } = this.state;
    return (
      <OrdersPageContext.Provider value={{ handleShowDetail: this.handleShowDetail, handleRmoveUnloginOrder: this.handleToggleRemoveAlertDialog, handleLinkOrder: this.handleToggleLinkAlertDialog }}>
        {loginUserOrders.length !== 0 && loginUserOrderAmount !== null && (
          <OrderBrowser
            orders={loginUserOrders}
            title={`Orders for ${this.props.user.displayName}`}
            // onShowDetail={this.handleShowDetail}
            // onRemoveOrder={this.handleRmoveUnloginOrder}
            // onLinkOrder={this.handleLinkOrder}
            totalOrder={loginUserOrderAmount}
            offset={0}
            onPageNumberClick={this.fetchLoginUserOrders}
          />
        )}
        {unloginUserOrders.length !== 0 && (
          <OrderBrowser
            orders={unloginUserOrders}
            title="Unsaved Orders"
            // onShowDetail={this.handleShowDetail}
            // onRemoveOrder={this.handleRmoveUnloginOrder}
            // onLinkOrder={this.handleLinkOrder}
            totalOrder={tempOrderIds.length}
            offset={0}
            onPageNumberClick={this.fetchUnloginUserOrders}
          />
        )}
        <OrderDetailDialog order={currentOrder} open={isDetailedDialogOpen} onClose={this.toggleDetailDialog} menuItems={menuItems} />
        {tempOrderIds.length !== 0 && (
          <Fragment>
            <AlertDialog
              open={isRemoveAlertOpen}
              onClose={this.handleToggleRemoveAlertDialog}
              title="Remove this order from this device?"
              content="After remove this order, you will no longer see it in this device."
              onFirstButton={this.handleRmoveUnloginOrder}
              onSecondButton={this.handleToggleRemoveAlertDialog}
              firstButtonText="Remove"
              secondButtonText="Cancel"
            />
            {user._id && (
              <AlertDialog
                open={isLinkAlertOpen}
                onClose={this.handleToggleLinkAlertDialog}
                title="Is it your order?"
                content="Link this order to your account."
                onFirstButton={this.handleLinkOrder}
                onSecondButton={this.handleToggleLinkAlertDialog}
                firstButtonText="Link"
                secondButtonText="Cancel"
              />
            )}
            <CheckCircleIconSnackbar open={isSnackbarOpen} onClose={this.handleToggleSnackbar} message={snackbarMessage} />
          </Fragment>
        )}
      </OrdersPageContext.Provider>
    );
  }
}
/* istanbul ignore next */
const mapStateToProps = state => ({
  orders: state.orders,
  user: state.user,
  tempOrderIds: state.tempOrderIds,
  menuItems: state.menuItems,
  loginUserOrderAmount: state.orders.loginUserOrderAmount
});
/* istanbul ignore next */
const mapDispatchToProps = dispatch => ({
  fetchOrderAmount: jwtMessage => dispatch(fetchOrderAmount(jwtMessage)),
  fetchAllMenu: () => dispatch(fetchAllMenu()),
  removeTempOrderId: orderId => dispatch(removeTempOrderId(orderId)),
  linkOrderToAccount: (orderId, jwt) => dispatch(linkOrderToAccount(orderId, jwt)),
});
export default connect(mapStateToProps, mapDispatchToProps)(OrdersPageContainer);
