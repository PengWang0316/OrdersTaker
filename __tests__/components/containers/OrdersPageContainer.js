import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';

import { OrdersPageContainer } from '../../../app/components/containers/OrdersPageContainer';

jest.mock('../../../app/components/OrderBrowser', () => 'OrderBrowser');
jest.mock('../../../app/components/OrderDetailDialog', () => 'OrderDetailDialog');
jest.mock('../../../app/components/AlertDialog', () => 'AlertDialog');
jest.mock('../../../app/components/snackbars/CheckCircleIconSnackbar', () => 'CheckCircleIconSnackbar');
jest.mock('../../../app/actions/OrdersActions', () => ({
  // fetchOrderAmount: jest.fn().mockReturnValue(Promise.resolve(10)),
  fetchUnloginUserOrders: jest.fn().mockReturnValue(Promise.resolve([])),
  fetchLoginUserOrders: jest.fn().mockReturnValue(Promise.resolve([]))
}));
jest.mock('../../../app/actions/TempOrderIdsActions', () => ({ linkOrderToAccount: jest.fn() }));

describe('OrdersPageContainer', () => {
  const defaultProps = {
    user: {
      _id: 'id',
      jwt: 'jwt'
    },
    orders: {
      loginUserOrderAmount: 7
    },
    tempOrderIds: [1, 2],
    menuItems: {
      1: {},
      2: {}
    },
    fetchOrderAmount: jest.fn(),
    fetchAllMenu: jest.fn(),
    loginUserOrderAmount: 2,
    removeTempOrderId: jest.fn()
  };
  const getShallowComponent = (props = defaultProps) => shallow(<OrdersPageContainer {...props} />);

  test('Initial states, constructor, and componentDidMount', () => {
    const OrdersActions = require('../../../app/actions/OrdersActions');
    let component = getShallowComponent({ ...defaultProps, user: {}, tempOrderIds: [] });
    expect(component.state('isDetailedDialogOpen')).toBe(false);
    expect(component.state('isRemoveAlertOpen')).toBe(false);
    expect(component.state('isLinkAlertOpen')).toBe(false);
    expect(component.state('isSnackbarOpen')).toBe(false);
    expect(component.state('snackbarMessage')).toBe('');
    expect(component.state('loginUserOrders').length).toBe(0);
    expect(component.state('unloginUserOrders').length).toBe(0);
    expect(defaultProps.fetchAllMenu).not.toHaveBeenCalled();
    expect(OrdersActions.fetchLoginUserOrders).not.toHaveBeenCalled();
    expect(OrdersActions.fetchUnloginUserOrders).not.toHaveBeenCalled();
    expect(defaultProps.fetchOrderAmount).not.toHaveBeenCalled();

    component = getShallowComponent({ ...defaultProps, menuItems: {}, user: {}, tempOrderIds: [] });
    expect(defaultProps.fetchAllMenu).toHaveBeenCalledTimes(1);
    expect(OrdersActions.fetchLoginUserOrders).not.toHaveBeenCalled();
    expect(OrdersActions.fetchUnloginUserOrders).not.toHaveBeenCalled();
    expect(defaultProps.fetchOrderAmount).not.toHaveBeenCalled();

    component = getShallowComponent();
    expect(OrdersActions.fetchLoginUserOrders).toHaveBeenCalledTimes(1);
    expect(OrdersActions.fetchLoginUserOrders).toHaveBeenLastCalledWith(0, { _id: 'id', jwt: 'jwt' });
    expect(OrdersActions.fetchUnloginUserOrders).toHaveBeenCalledTimes(1);
    expect(OrdersActions.fetchUnloginUserOrders).toHaveBeenLastCalledWith(0, [1, 2]);
    expect(defaultProps.fetchOrderAmount).not.toHaveBeenCalled();

    component = getShallowComponent({ ...defaultProps, orders: { loginUserOrderAmount: null } });
    expect(defaultProps.fetchOrderAmount).toHaveBeenCalledTimes(1);
    expect(defaultProps.fetchOrderAmount).toHaveBeenLastCalledWith(defaultProps.user);
  });

  test('getNewUnloginOrdersArray', () =>
    expect(OrdersPageContainer.getNewUnloginOrdersArray([{ _id: 'id1' }, { _id: 'id2' }, { _id: 'id3' }], 'id2'))
      .toEqual([{ _id: 'id1' }, { _id: 'id3' }]));

  test('fetchLoginUserOrders without error', async () => {
    console.error = jest.fn();
    const OrdersActions = require('../../../app/actions/OrdersActions');
    OrdersActions.fetchLoginUserOrders.mockReturnValue(Promise.resolve([{ _id: 'orderId' }]));
    const component = getShallowComponent();
    await component.instance().fetchLoginUserOrders(10);
    expect(OrdersActions.fetchLoginUserOrders).toHaveBeenLastCalledWith(10, { _id: 'id', jwt: 'jwt' });
    expect(component.state('loginUserOrders')).toEqual([{ _id: 'orderId' }]);
    expect(console.error).not.toHaveBeenCalled();
  });

  test('fetchLoginUserOrders with error', async () => {
    window.console.error = jest.fn();
    const OrdersActions = require('../../../app/actions/OrdersActions');
    OrdersActions.fetchLoginUserOrders.mockReturnValue(Promise.reject());
    const component = getShallowComponent();
    await component.instance().fetchLoginUserOrders(10);
    expect(OrdersActions.fetchLoginUserOrders).toHaveBeenLastCalledWith(10, { _id: 'id', jwt: 'jwt' });
    expect(component.state('loginUserOrders')).toEqual([]);
    expect(window.console.error).toHaveBeenCalled();
  });

  test('fetchUnloginUserOrders without error', async () => {
    const OrdersActions = require('../../../app/actions/OrdersActions');
    OrdersActions.fetchUnloginUserOrders.mockReturnValue(Promise.resolve([{ _id: 'orderId' }]));
    const component = getShallowComponent();
    await component.instance().fetchUnloginUserOrders(10);
    expect(OrdersActions.fetchUnloginUserOrders).toHaveBeenLastCalledWith(10, [1, 2]);
    expect(component.state('unloginUserOrders')).toEqual([{ _id: 'orderId' }]);
    // expect(console.error).not.toHaveBeenCalled();
  });

  test('fetchUnloginUserOrders with error', async () => {
    window.console.error = jest.fn();
    const OrdersActions = require('../../../app/actions/OrdersActions');
    OrdersActions.fetchUnloginUserOrders.mockReturnValue(Promise.reject());
    const component = getShallowComponent();
    await component.instance().fetchUnloginUserOrders(10);
    expect(OrdersActions.fetchUnloginUserOrders).toHaveBeenLastCalledWith(10, [1, 2]);
    expect(component.state('unloginUserOrders')).toEqual([]);
    expect(window.console.error).toHaveBeenCalled();
  });

  test('toggleDetailDialog', () => {
    const component = getShallowComponent();
    expect(component.state('isDetailedDialogOpen')).toBe(false);
    component.instance().toggleDetailDialog();
    expect(component.state('isDetailedDialogOpen')).toBe(true);
  });

  test('handleShowDetail', () => {
    const component = getShallowComponent();
    const order = { _id: 'orderId' };
    expect(component.state('currentOrder')).toBeUndefined();
    component.instance().handleShowDetail(order);
    expect(component.state('currentOrder')).toBe(order);
  });

  test('handleRmoveUnloginOrder', async () => {
    const component = getShallowComponent();
    component.instance().unloginOrderId = 'unloginOrderId';
    const mockHandleToggleRemoveAlertDialog = jest.fn();
    const mockHandleToggleSnackbar = jest.fn();
    component.instance().handleToggleRemoveAlertDialog = mockHandleToggleRemoveAlertDialog;
    component.instance().handleToggleSnackbar = mockHandleToggleSnackbar;
    component.setState({ unloginUserOrders: [{ _id: '1' }, { _id: 'unloginOrderId' }] });

    await component.instance().handleRmoveUnloginOrder();
    expect(defaultProps.removeTempOrderId).toHaveBeenLastCalledWith('unloginOrderId');
    expect(mockHandleToggleRemoveAlertDialog).toHaveBeenCalledTimes(1);
    expect(mockHandleToggleRemoveAlertDialog).toHaveBeenLastCalledWith(null);
    expect(mockHandleToggleSnackbar).toHaveBeenCalledTimes(1);
    expect(component.state('unloginUserOrders')).toEqual([{ _id: '1' }]);
    expect(component.state('snackbarMessage')).toEqual('Removed Successfully');
  });

  test('handleLinkOrder', async () => {
    const TempOrderIdsActions = require('../../../app/actions/TempOrderIdsActions');
    const component = getShallowComponent();
    component.instance().unloginOrderId = 'unloginOrderId';
    const mockHandleToggleLinkAlertDialog = jest.fn();
    const mockHandleToggleSnackbar = jest.fn();
    component.instance().handleToggleLinkAlertDialog = mockHandleToggleLinkAlertDialog;
    component.instance().handleToggleSnackbar = mockHandleToggleSnackbar;
    component.setState({ unloginUserOrders: [{ _id: '1' }, { _id: 'unloginOrderId' }] });

    await component.instance().handleLinkOrder();
    expect(defaultProps.removeTempOrderId).toHaveBeenLastCalledWith('unloginOrderId');
    expect(mockHandleToggleLinkAlertDialog).toHaveBeenCalledTimes(1);
    expect(mockHandleToggleLinkAlertDialog).toHaveBeenLastCalledWith(null);
    expect(mockHandleToggleSnackbar).toHaveBeenCalledTimes(1);
    expect(component.state('unloginUserOrders')).toEqual([{ _id: '1' }]);
    expect(component.state('snackbarMessage')).toEqual('Linked to your account sccessfully');
    expect(TempOrderIdsActions.linkOrderToAccount).toHaveBeenCalledTimes(1);
    expect(TempOrderIdsActions.linkOrderToAccount).toHaveBeenLastCalledWith('unloginOrderId', 'jwt');
  });

  test('handleToggleRemoveAlertDialog', async () => {
    const component = getShallowComponent();
    expect(component.state('isRemoveAlertOpen')).toBe(false);

    await component.instance().handleToggleRemoveAlertDialog('orderId');
    expect(component.instance().unloginOrderId).toEqual('orderId');
    expect(component.state('isRemoveAlertOpen')).toBe(true);
  });

  test('handleToggleLinkAlertDialog', async () => {
    const component = getShallowComponent();
    expect(component.state('isLinkAlertOpen')).toBe(false);

    await component.instance().handleToggleLinkAlertDialog('orderId');
    expect(component.instance().unloginOrderId).toEqual('orderId');
    expect(component.state('isLinkAlertOpen')).toBe(true);
  });

  test('handleToggleSnackbar', async () => {
    const component = getShallowComponent();
    expect(component.state('isSnackbarOpen')).toBe(false);

    await component.instance().handleToggleSnackbar();
    expect(component.state('isSnackbarOpen')).toBe(true);
  });

  test('Snapshot', () => {
    const component = getShallowComponent();
    component.setState({ loginUserOrders: [{}], unloginUserOrders: [{}] });
    expect(renderer.create(component).toJSON()).toMatchSnapshot();
  });
});
