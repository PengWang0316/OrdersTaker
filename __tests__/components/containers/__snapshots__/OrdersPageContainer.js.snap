// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrdersPageContainer Snapshot 1`] = `
Array [
  <OrderBrowser
    offset={0}
    onPageNumberClick={[Function]}
    orders={
      Array [
        Object {},
      ]
    }
    title="Orders for undefined"
    totalOrder={2}
  />,
  <OrderBrowser
    offset={0}
    onPageNumberClick={[Function]}
    orders={
      Array [
        Object {},
      ]
    }
    title="Unsaved Orders"
    totalOrder={2}
  />,
  <OrderDetailDialog
    menuItems={
      Object {
        "1": Object {},
        "2": Object {},
      }
    }
    onClose={[Function]}
    open={false}
  />,
  <AlertDialog
    content="After remove this order, you will no longer see it in this device."
    firstButtonText="Remove"
    onClose={[Function]}
    onFirstButton={[Function]}
    onSecondButton={[Function]}
    open={false}
    secondButtonText="Cancel"
    title="Remove this order from this device?"
  />,
  <AlertDialog
    content="Link this order to your account."
    firstButtonText="Link"
    onClose={[Function]}
    onFirstButton={[Function]}
    onSecondButton={[Function]}
    open={false}
    secondButtonText="Cancel"
    title="Is it your order?"
  />,
  <CheckCircleIconSnackbar
    message=""
    onClose={[Function]}
    open={false}
  />,
]
`;
